File: AddTwoNumbersAsLists/AddTwoNumbersAsLists.cpp
// AddTwoNumbersAsLists.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

void PrintList(ListNode* pHead)
{
    if (!pHead)
        return;
    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }

    std::cout << std::endl;
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

std::vector<int> GetArrayFromInput(int count, const std::string& prompt)
{
    if (count == 0)
        return std::vector<int>();

    std::cout << prompt;
    std::vector<int> vec(count, 0);
    for (int i = 0; i < count; ++i)
        std::cin >> vec[i];

    return vec;
}

ListNode* GetListFromArray(const std::vector<int>& vec)
{
    if (vec.size() == 0)
        return nullptr;

    ListNode* pHead = new ListNode(vec[0]);
    ListNode* pCur = pHead;
    for (size_t i = 1; i < vec.size(); ++i)
    {
        pCur->next = new ListNode(vec[i]);
        pCur = pCur->next;
    }

    return pHead;
}

ListNode* AddNode(ListNode* resultNode, ListNode* pFirstNode, ListNode* pSecondNode)
{
    if (pFirstNode == nullptr && pSecondNode == nullptr)
    {
        return resultNode;
    }

    int sum = 0;
    if (pFirstNode != nullptr)
    {
        sum += pFirstNode->data;
        pFirstNode = pFirstNode->next;
    }
    if (pSecondNode != nullptr)
    {
        sum += pSecondNode->data;
        pSecondNode = pSecondNode->next;
    }
    if (resultNode != nullptr)
    {
        sum += resultNode->data;
    }

    ListNode* pNewNode = new ListNode(sum % 10);
    if (resultNode != nullptr)
    {
        delete resultNode;
        resultNode = nullptr;
    }

    if (sum / 10 != 0)
    {
        resultNode = new ListNode(sum / 10);
    }

    pNewNode->next = AddNode(resultNode, pFirstNode, pSecondNode);
    return pNewNode;
}

ListNode* AddTwoNumbers(ListNode* firstList, ListNode* secondList)
{
    ListNode* pResult = nullptr;
    return AddNode(pResult, firstList, secondList);
}

int main()
{
    while (true)
    {
        int firstNumber = 0;
        int secondNumber = 0;
        std::cout << "Enter the two numbers: ";
        std::cin >> firstNumber >> secondNumber;

        if (firstNumber == 0 && secondNumber == 0)
            break;

        std::vector<int> vec = GetArrayFromInput(firstNumber, "Please enter the first list: ");
        ListNode* pFirstHead = GetListFromArray(vec);

        vec = GetArrayFromInput(secondNumber, "Please enter the second list: ");
        ListNode* pSecondHead = GetListFromArray(vec);

        ListNode* pSumHead = AddTwoNumbers(pFirstHead, pSecondHead);

        std::cout << "The two lists to be added: " << std::endl;
        PrintList(pFirstHead);
        PrintList(pSecondHead);
        std::cout << "The sum of the two lists: " << std::endl;
        PrintList(pSumHead);

        FreeList(pFirstHead);
        FreeList(pSecondHead);
        FreeList(pSumHead);
    }

    return 0;
}

File: Go/AddTwoNumbersAsLists.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) *ListNode {
	if len(v) == 0 {
		return nil
	}

	if len(v) == 1 {
		return &ListNode{v[0], nil}
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func AddNodes(pResult, first, second *ListNode) *ListNode {
	if first == nil && second == nil {
		return pResult
	}

	sum := 0
	if pResult != nil {
		sum += pResult.data
	}
	if first != nil {
		sum += first.data
	}
	if second != nil {
		sum += second.data
	}

	pNewNode := &ListNode{sum % 10, nil}
	if sum/10 != 0 {
		pResult = &ListNode{sum / 10, nil}
	} else {
		pResult = nil
	}
	if first != nil {
		first = first.next
	}
	if second != nil {
		second = second.next
	}

	pNewNode.next = AddNodes(pResult, first, second)
	return pNewNode
}

func AddTwoNumbers(firstList, secondList *ListNode) *ListNode {
	var pResult *ListNode = nil
	return AddNodes(pResult, firstList, secondList)
}

func main() {
	for {
		fmt.Print("Number of elements of list1 and list2: ")
		var count1, count2 int
		fmt.Scan(&count1, &count2)
		if count1 == 0 && count2 == 0 {
			break
		}

		vec := GetArrayFromInput(count1, "Please enter the first list: ")
		first := GetListFromArray(vec)

		vec = GetArrayFromInput(count2, "Please enter the second list: ")
		second := GetListFromArray(vec)

		result := AddTwoNumbers(first, second)

		fmt.Print("The sum of the two lists is ")
		result.Print()
	}
}
